model-optimizer/extensions/front/tf/KerasRNNReplacer.py

```diff
@@ -0,0 +1,237 @@
+from mo.graph.graph import Graph, Node, rename_nodes
+from mo.middle.pattern_match import find_pattern_matches, inverse_dict
+from mo.ops.const import Const
+from mo.ops.squeeze import Squeeze
+from mo.ops.unsqueeze import Unsqueeze
+
+
+def find_intergraph_match(graph: Graph, ext_match: dict, reserve_port: int, unstack_port: int, stack_port: int):
+    """
+    The function checks that logic using TensorList operations corresponds to subsequent slicing of input and
+    concatenation of intermediate results for output.
+    If pattern is successfully found, it returns matching dictionary.
+    """
+    def back_edge_exists(back_edges_map, from_layer, to_layer, from_port=0, to_port=0):
+        for back_edge in back_edges_map:
+            if back_edge['from_layer'] == from_layer and back_edge['to_layer'] == to_layer \
+                    and back_edge['from_port'] == from_port and back_edge['to_port'] == to_port:
+                return True
+        return False
+
+    def inter_edge_exists(input_port_map, external_port_id, internal_layer_id):
+        for input_port in input_port_map:
+            if input_port['external_port_id'] == external_port_id and \
+                    input_port['internal_layer_id'] == internal_layer_id:
+                return True
+        return False
+
+    loop_node = ext_match['while']
+    body_graph = loop_node['body']
+
+    # check that a list of tensors are iterated subsequently in the body graph
+    # that is functionally equivalent to slicing a tensor through a certain dimension
+    body_pattern = dict(
+        nodes=[('container', dict(op='Parameter')),
+               ('tensor_list', dict(op='Parameter')),
+               ('current_iteration', dict(op='Parameter')),
+               ('slicing', dict(op='TensorListGetItem')),
+               ('const_increment', dict(op='Const')),
+               ('increment_iteration', dict(op='Add')),
+               ('increment_iteration_identity', dict(op='Identity')),
+               ('increment_iteration_result', dict(op='Result')),
+               ('concatenation', dict(op='TensorListSetItem')),
+               ('concatenation_identity', dict(op='Identity')),
+               ('concatenation_result', dict(op='Result')),
+               ],
+        edges=[('tensor_list', 'slicing', {'in': 0}),
+               ('current_iteration', 'slicing', {'in': 1}),
+               ('const_increment', 'increment_iteration', {'in': 1}),
+               ('current_iteration', 'increment_iteration', {'in': 0}),
+               ('container', 'concatenation', {'in': 0}),
+               ('current_iteration', 'concatenation', {'in': 1}),
+               ('concatenation', 'concatenation_identity', {'in': 0}),
+               ('concatenation_identity', 'concatenation_result', {'in': 0}),
+               ('increment_iteration', 'increment_iteration_identity', {'in': 0}),
+               ('increment_iteration_identity', 'increment_iteration_result', {'in': 0}),
+               ]
+    )
```

Comment:
Do we really want to match both inputs and outputs in one pattern? May be it make sense to match them separately, for, example, there could be a case when the input is sliced but the output is not concatenated? Can you provide pros/cons for each of the approaches?

Reply:
Done. I did split into a couple of transformation.